Sofa/framework/Simulation/Core/src/sofa/simulation/DefaultAnimationLoop.cpp

```diff
@@ -33,122 +33,329 @@
+void DefaultAnimationLoop::propagateCollisionEndEvent(const core::ExecParams* params) const
+{
+    CollisionEndEvent evEnd;
+    PropagateEventVisitor eventPropagation( params, &evEnd);
+    eventPropagation.execute(gnode);
+}
 
+void DefaultAnimationLoop::collisionDetection(const core::ExecParams* params) const
+{
+    propagateCollisionBeginEvent(params);
 
-    sofa::helper::AdvancedTimer::stepBegin("AnimateVisitor");
-    AnimateVisitor act(params, dt);
-    gnode->execute ( act );
-    sofa::helper::AdvancedTimer::stepEnd("AnimateVisitor");
+    {
+        sofa::helper::ScopedAdvancedTimer timer("collision");
+        CollisionVisitor act(params);
+        gnode->execute(&act);
+    }
 
+    propagateCollisionEndEvent(params);
+}
+
+void DefaultAnimationLoop::computeIsolatedForces(const core::ExecParams* params, SReal dt) const
+{
+    sofa::helper::ScopedAdvancedTimer timer("computeIsolatedForces");
+    ComputeIsolatedForceVisitor visitor(params, dt);
+    gnode->execute(&visitor);
+}
+
+void DefaultAnimationLoop::animate(const core::ExecParams* params, SReal dt) const
+{
+    const SReal startTime = gnode->getTime();
+    const SReal nextTime = startTime + dt;
+
+    sofa::core::MechanicalParams mparams(*params);
+    mparams.setDt(dt);
 
-    sofa::helper::AdvancedTimer::stepBegin("UpdateSimulationContextVisitor");
-    gnode->setTime ( startTime + dt );
-    gnode->execute< UpdateSimulationContextVisitor >(params);
-    sofa::helper::AdvancedTimer::stepEnd("UpdateSimulationContextVisitor");
+    behaviorUpdatePosition(params, dt);
+    updateInternalData(params);
 
+    resetConstraint(params);
+
+    beginIntegration(params, dt);
     {
-        AnimateEndEvent ev ( dt );
-        PropagateEventVisitor propagateEventVisitor ( params, &ev );
-        gnode->execute ( propagateEventVisitor );
+        collisionDetection(params);
+
+        const core::ConstraintParams cparams;
+        buildConstraintMatrix(cparams);
+        accumulateMatrixDeriv(cparams);
+
+        solve(params, dt);
+
+        projectPositionAndVelocity(nextTime, mparams);
+        propagateOnlyPositionAndVelocity(nextTime, mparams);
     }
+    endIntegration(params, dt);
+    computeIsolatedForces(params, dt);
 
-    sofa::helper::AdvancedTimer::stepBegin("UpdateMapping");
-    //Visual Information update: Ray Pick add a MechanicalMapping used as VisualMapping
-    gnode->execute< UpdateMappingVisitor >(params);
+    updateSimulationContext(params, dt, startTime);
+}
+
+void DefaultAnimationLoop::step(const core::ExecParams* params, SReal dt)
+{
+    if (this->d_componentState.getValue() != sofa::core::objectmodel::ComponentState::Valid)
     {
-        UpdateMappingEndEvent ev ( dt );
-        PropagateEventVisitor propagateEventVisitor ( params , &ev );
-        gnode->execute ( propagateEventVisitor );
+        return;
     }
-    sofa::helper::AdvancedTimer::stepEnd("UpdateMapping");
 
-    if (d_computeBoundingBox.getValue())
+    if (dt == 0_sreal)
```

Comment:
I know it was there before but this look weird to me. The function is overriding its argument if dt is zero. I think a better behavior is do nothing if dt == 0 (or warn user) instead of picking the value from the node.

Reply:
Do you mind if we do it in another PR? The thing is, "all" animation loops do that